unpublished/build-system-abstraction-njs.rst

```diff
@@ -0,0 +1,708 @@
+new versions of e.g. ``pip`` and ``flit`` are both updated to support
+the new interface, then this should be sufficient for it to be used;
+in particular, it should *not* be necessary for every project that
+*uses* ``flit`` to update its individual ``pypackage.json`` file. (b)
+We do not want to have to spawn extra processes just to perform this
+negotation, because process spawns can easily become a bottleneck when
+deploying large multi-package stacks on some platforms (Windows).
+
+In the interface described here, all of these goals are easy to
+achieve. Because ``pip`` controls the code that runs inside the child
+process, it can easily write it to do something like::
+
+    command, backend, args = parse_command_line_args(...)
+    if command == "do_editable_install":
+       if hasattr(backend, "install_editable2"):
+           backend.install_editable2(...)
+       elif hasattr(backend, "install_editable"):
+           backend.install_editable(...)
+       else:
+           # error handling
+
+In the alternative where the public interface boundary is placed at
+the subprocess call, this is not possible -- either we need to spawn
+an extra process just to query what interfaces are supported (as was
+included in an earlier version of `this alternative PEP
+<https://github.com/pypa/interoperability-peps/pull/54/files>`_), or
+else we give up on autonegotiation entirely (as in the current version
+of that PEP), meaning that any changes in the interface will require
+N individual packages to update their ``pypackage.json`` files before
+any change can go live, and that any changes will necessarily be
+restricted to new releases.
+
+One specific consequence of this is that in this PEP, we're able to
+make the ``get_wheel_metadata`` command optional. In our design, this
+can easily be worked around by a tool like ``pip``, which can put code
+in its subprocess runner like::
+
+    def get_wheel_metadata(output_dir, config_settings):
+         if hasattr(backend, "get_wheel_metadata"):
+             backend.get_wheel_metadata(output_dir, config_settings)
+         else:
+             backend.build_wheel(output_dir, config_settings)
+             touch(output_dir / "PIP_ALREADY_BUILT_WHEELS")
+             unzip_metadata(output_dir/*.whl)
+
+    def build_wheel(output_dir, config_settings, metadata_dir):
+         if os.path.exists(metadata_dir / "PIP_ALREADY_BUILT_WHEELS"):
+             copy(metadata_dir / *.whl, output_dir)
+         else:
+             backend.build_wheel(output_dir, config_settings, metadata_dir)
+
+and thus expose a totally uniform interface to the rest of ``pip``,
+with no extra subprocess calls, no duplicated builds, etc. But
+obviously this is the kind of code that you only want to write as part
+of a private, within-project interface.
+
+(And, of course, making the ``metadata`` command optional is one piece
+of lowering the barrier to entry, as discussed above.)
+
+
```

Comment:
One downside of a Python-based interface is that it makes it harder to write build systems that aren't implemented in Python. I'm not saying this is necessarily a good thing to do, but in the spirit of your comment above about quick hacks developing into significant tools, I could imagine someone wanting to write a primitive build tool around something like make and a few shell scripts.

Of course you can't specify "I need make and bash" as a PEP 508 style requirement, so in practice the fact that you have to be able to indicate the build system using ```bootstrap_requires``` precludes such options anyway.